ledger/participant-integration-api/src/main/scala/platform/store/dao/EventProjectionProperties.scala

```diff
@@ -0,0 +1,99 @@
+  * @param populateInterfaceView populate interface_views. The Map of templates to interfaces,
+  *                              and the set of implementor templates cannot be empty.
+  */
+final case class EventProjectionProperties private[dao] (
+    verbose: Boolean,
+    // Map(eventWitnessParty, Set(templateId))
+    populateContractArgument: Map[String, Set[Identifier]] = Map.empty,
+    // Map(eventWitnessParty, Map(templateId -> Set(interfaceId)))
+    populateInterfaceView: Map[String, Map[Identifier, Set[Identifier]]] = Map.empty,
+) {
+  def render(witnesses: Seq[String], templateId: Identifier): RenderResult = {
+    val renderContractArguments = witnesses
+      .map(populateContractArgument.get)
+      .exists {
+        case Some(wildcardTemplates) if wildcardTemplates.isEmpty => true
+        case Some(nonEmptyTemplates) if nonEmptyTemplates(templateId) => true
+        case _ => false
+      }
+
+    val renderInterfaces: Seq[Identifier] =
+      (for {
+        witness <- witnesses.iterator
+        templateToInterfaceMap <- populateInterfaceView.get(witness).toList
+        interfaces <- templateToInterfaceMap.getOrElse(templateId, Set.empty[Identifier])
+      } yield interfaces).toSeq
+
+    RenderResult(renderContractArguments, renderInterfaces)
+  }
+
+}
+
+object EventProjectionProperties {
+
+  case class RenderResult(
+      contractArguments: Boolean,
+      interfaces: Seq[Identifier],
+  )
+
+  def apply(
+      domainTransactionFilter: domain.TransactionFilter,
+      verbose: Boolean,
+      interfaceImplementedBy: Ref.Identifier => Set[Ref.Identifier],
+  ): EventProjectionProperties =
+    EventProjectionProperties(
+      verbose = verbose,
+      populateContractArgument = populateContractArgument(domainTransactionFilter),
+      populateInterfaceView = populateInterfaceView(domainTransactionFilter, interfaceImplementedBy),
+    )
+
+  private def populateContractArgument(domainTransactionFilter: domain.TransactionFilter) = {
+    (for {
+      (party, filters) <- domainTransactionFilter.filtersByParty.iterator
+      inclusiveFilters <- filters.inclusive.iterator
+      templateId <- inclusiveFilters.templateIds.iterator
+    } yield party.toString -> templateId)
+      .toSet[(String, Identifier)]
+      .groupMap(_._1)(_._2)
+      .++(
+        domainTransactionFilter.filtersByParty.iterator
+          .collect {
+            case (party, Filters(None)) =>
+              party.toString -> Set.empty[Identifier]
+
+            case (party, Filters(Some(empty)))
+                if empty.templateIds.isEmpty && empty.interfaceFilters.isEmpty =>
+              party.toString -> Set.empty[Identifier]
+          }
+      )
+  }
+
+  private def populateInterfaceView(
+      domainTransactionFilter: domain.TransactionFilter,
+      interfaceImplementedBy: Identifier => Set[Ref.Identifier],
+  ) = (for {
```

Comment:
how about adding return type for more readability?

Reply:
Im not always adding a return type to private defs :)

Reply:
I would consider adding it in non trivial cases, since it is much more readable that way (at least for me :thinking: )

Reply:
depends what do you use to review the code, Intellij shows the type anyway ;) but let's make it more readable, I agree here, the type is non trivial one